### PR TITLE
 Make model reshape and track batch (#12736)

### PR DESCRIPTION
**Pull request by jane-intel**
Cherry pick from #12736 and resolved comments

### CODE REVIEW

File: src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+#include "openvino/opsets/opset9.hpp"
+#include "openvino/pass/manager.hpp"
+#include "transformations/utils/utils.hpp"
+
+using namespace std;
+using namespace ov::opset9;
+
+ov::Input<ov::Node> get_outer_input_of_ti_by_parameter(const shared_ptr<Parameter>& parameter,
+                                                       const shared_ptr<TensorIterator>& ti) {
+    const auto& body = ti->get_body();  // body is not nullptr -- we checked earlier
```
Comment (jane-intel) [inline src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp:21]:
```suggestion
```

File: src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                                            Constant::create(ov::element::i64, ov::Shape{}, {0}));
+    return batch;
+}
+
+bool broadcast_state_by_batch(ov::Input<ov::Node> input, const shared_ptr<ov::Node>& batch_delivering_node) {
+    auto constant_state = dynamic_pointer_cast<Constant>(input.get_source_output().get_node_shared_ptr());
+    if (constant_state == nullptr)
+        return false;
+    const auto& constant_shape = constant_state->get_shape();
+    if (constant_shape[0] != 1)
```
Comment (itikhono) [inline src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp:102]: do we need to check rank here? maybe constant is scalar

File: src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    rewritten |= broadcast_state_by_batch(lstm_cell->input(2), batch_delivering_node);
+    return rewritten;
+}
+
+bool ov::pass::LSTMStatesBroadcast::run_on_model(const shared_ptr<ov::Model>& f) {
+    RUN_ON_FUNCTION_SCOPE(LSTMStatesBroadcast);
+    bool rewritten = false;
+    for (auto& node : f->get_ordered_ops()) {
+        // Recursively apply transformation for sub-graph based operations
+        if (const auto& sub_graph_node = dynamic_pointer_cast<ov::op::util::SubGraphOp>(node))
```
Comment (itikhono) [inline src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp:152]: SubGraphOp -> MultiSubgraphOp, e.g. If op
We also need to run this transformation for all bodies

File: src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp (hunk 2 above)
Comment (jane-intel) [reply src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp:102]: It couldn't. We believe network is validated before and it goes into LSTMCell, so there is no need to check.

File: src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp (hunk 3 above)
Comment (jane-intel) [reply src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp:152]: Let's leave it for the time we have a real model with such case.

File: src/common/transformations/src/transformations/smart_reshape/reshape_sinking.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            return false;
+        auto output_pattern = constant->cast_vector<int64_t>();
+        if (output_pattern.size() != input_rank)
+            return false;
+        if (!all_of(output_pattern.begin(), output_pattern.end(), [](const int64_t& i) {
+                return i > 0;
+            }))
+            return false;
+        for (size_t i = 1; i < input_rank; ++i) {
+            if (i + 1 == input_rank) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/smart_reshape/reshape_sinking.cpp:120]: nitpick: use output_patter[input_rank] != O, not nessary to check it in the loop

File: src/common/transformations/src/transformations/smart_reshape/reshape_sinking.cpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            if (!add || add->get_autob() != ov::op::AutoBroadcastType::NUMPY)
+                return false;
+            const auto& constant = dynamic_pointer_cast<Constant>(add->get_input_node_shared_ptr(1));
+            if (!constant)
+                return false;
+            const auto& constant_shape = constant->get_shape();
+            auto desired_ones_shape = ov::Shape(constant_shape.size(), 1);
+            auto desired_shape = ov::Shape(constant_shape.size() - 1, 1);
+            desired_shape.push_back(O);
+            OPENVINO_ASSERT(constant_shape.size() == desired_ones_shape.size() &&
```
Comment (itikhono) [inline src/common/transformations/src/transformations/smart_reshape/reshape_sinking.cpp:99]: nitpick: it looks like it never throw, we can delete it

**Review by itikhono — APPROVED** [review]
LGTM

**Pull request merged**